File: cpp/13_polymorphism.cpp
// ============================================================
// Topic:  Polymorphism
// Summary: Treat different types through a base pointer; the right method runs.
// Run:     g++ -std=c++17 cpp/13_polymorphism.cpp -o cpp/13_polymorphism && cpp/13_polymorphism
// ============================================================

#include <iostream>
#include <iomanip>
#include <vector>
#include <memory>
#include <string>

const double PI = 3.141592653589793;

class Shape {
public:
    virtual ~Shape() = default;
    virtual std::string name() const { return "Shape"; }
    virtual double area() const { return 0.0; }
};

class Circle : public Shape {
    double r;
public:
    Circle(double r) : r(r) {}
    std::string name() const override { return "Circle"; }
    double area() const override { return PI * r * r; }
};

class Rectangle : public Shape {
    double w, h;
public:
    Rectangle(double w, double h) : w(w), h(h) {}
    std::string name() const override { return "Rectangle"; }
    double area() const override { return w * h; }
};

int main() {
    // Polymorphism needs a pointer or reference to the base type.
    std::vector<std::unique_ptr<Shape>> shapes;
    shapes.push_back(std::make_unique<Circle>(2));
    shapes.push_back(std::make_unique<Rectangle>(3, 4));

    double total = 0;
    std::cout << std::fixed << std::setprecision(2);
    for (const auto& s : shapes) {
        std::cout << s->name() << " area: " << s->area() << std::endl;
        total += s->area();
    }
    std::cout << "Total area: " << total << std::endl;
    // Output:
    // Circle area: 12.57
    // Rectangle area: 12.00
    // Total area: 24.57
    return 0;
}

// Cross-language note:
//   Dynamic dispatch in C++ happens only through a base POINTER or REFERENCE to a
//   `virtual` method. Storing objects by value (a std::vector<Shape>) would "slice"
//   them down to the base, so we hold unique_ptr<Shape> instead.

File: rust/13_polymorphism.rs
// ============================================================
// Topic:  Polymorphism  (Rust: trait objects — Box<dyn Trait>)
// Summary: Store different types behind one trait; calls dispatch dynamically.
// Run:     rustc rust/13_polymorphism.rs -o rust/13_polymorphism && rust/13_polymorphism
// ============================================================

trait Shape {
    fn name(&self) -> String;
    fn area(&self) -> f64;
}

struct Circle {
    r: f64,
}
impl Shape for Circle {
    fn name(&self) -> String { "Circle".to_string() }
    fn area(&self) -> f64 { std::f64::consts::PI * self.r * self.r }
}

struct Rectangle {
    w: f64,
    h: f64,
}
impl Shape for Rectangle {
    fn name(&self) -> String { "Rectangle".to_string() }
    fn area(&self) -> f64 { self.w * self.h }
}

fn main() {
    // Box<dyn Shape> = a "trait object": different concrete types, one interface.
    let shapes: Vec<Box<dyn Shape>> = vec![
        Box::new(Circle { r: 2.0 }),
        Box::new(Rectangle { w: 3.0, h: 4.0 }),
    ];

    let mut total = 0.0;
    for s in &shapes {
        println!("{} area: {:.2}", s.name(), s.area());
        total += s.area();
    }
    println!("Total area: {:.2}", total);
    // Output:
    // Circle area: 12.57
    // Rectangle area: 12.00
    // Total area: 24.57
}

// Cross-language note:
//   Rust has two flavors of polymorphism:
//     - STATIC (generics / `impl Trait`) — resolved at compile time, no runtime cost.
//     - DYNAMIC (`dyn Trait` behind a pointer like Box) — resolved at runtime via a
//       vtable, which is what lets a Vec hold a mix of Circle and Rectangle here.
